src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -357,6 +363,154 @@ Matrix1D<double> ProgSubtractProjection::checkBestModel(MultidimArray< std::comp
 	return R2;
 }
 
+void ProgSubtractProjection::generateNoiseEstimationSideInfo()
+{
+	// Initialize powerNoise
+	powerNoise.initZeros((int)Ydim, (int)Xdim/2 +1);
+
+	// Calculate boundaries for noise estimation 
+	// (make a more eficinet sampling of the subtracted region)
+	if(maskVolProvided)
+	{
+		int minX = Xdim;
+		int minY = Ydim;
+		int minZ = Zdim;
+		int maxX = 0;
+		int maxY = 0;
+		int maxZ = 0;
+
+		for(size_t k=0; k<Zdim; ++k)
+		{
+			for(size_t i=0; i<Ydim; ++i)
+			{
+				for(size_t j=0; j<Xdim; ++j)
+				{
+					if (DIRECT_A3D_ELEM(maskVol(), k, i, j) > 0) {
+						minX = std::min(minX, (int)i);
+						minY = std::min(minY, (int)j);
+						minZ = std::min(minZ, (int)k);
+						maxX = std::max(maxX, (int)i);
+						maxY = std::max(maxY, (int)j);
+						maxZ = std::max(maxZ, (int)k);
+					}
+				}
+			}
+		}
+
+		min_noiseEst = std::min(minX, std::min(minY, minZ));
+		max_noiseEst = std::max(maxX, std::max(maxY, maxZ));
+	}
+	else
+	{
+		// Assuming square particles
+		max_noiseEst = int((double)Xdim/2 + cirmaskrad/2);
+		min_noiseEst = int((double)Xdim/2 - cirmaskrad/2);
+	}
+
+	#ifdef DEBUG_NOISE_ESTIMATION
+	std::cout << "max_noiseEst  " << max_noiseEst << " min_noiseEst " << min_noiseEst  << std::endl;
+	#endif
+}
+
+void ProgSubtractProjection::noiseEstimation()
+{
+	#ifdef DEBUG_NOISE_ESTIMATION
+	std::cout << "Estimating noise from particle..." << std::endl;
+	#endif
+
+    srand(time(0)); // Seed for random number generation
+	double scallignFactor = (Xdim * Ydim) / (cropSize * cropSize);
+    bool invalidRegion;
+    MultidimArray< double > noiseCrop;
+
+	#ifdef DEBUG_NOISE_ESTIMATION
+	std::cout << "max_noiseEst  " << max_noiseEst << " min_noiseEst " << min_noiseEst  << std::endl;
+	std::cout << "(Ydim/2) " << (Ydim/2) << " (Xdim/2) " << (Xdim/2) << std::endl;
+	std::cout << "scallignFactor " << scallignFactor << std::endl;
+	#endif
+
+	do {
+		invalidRegion = false;
+		noiseCrop.initZeros((int)Ydim, (int)Xdim);
+
+		int x = min_noiseEst + rand() % (max_noiseEst - min_noiseEst + 1);
+		int y = min_noiseEst + rand() % (max_noiseEst - min_noiseEst + 1);
+
+		#ifdef DEBUG_NOISE_ESTIMATION
+		std::cout << "x  " << x << " y " << y  << std::endl;
+		#endif
+
+		for (size_t i = 0; i < cropSize; i++)
+		{
+			for (size_t j = 0; j < cropSize; j++)
+			{
+				#ifdef DEBUG_NOISE_ESTIMATION
+				std::cout << "--------------------------------------------------" << std::endl;
+				std::cout << "y + i  " << y + i << " x + j " << x + j << std::endl;
+				std::cout << "(Ydim/2) - (cropSize/2) + i  " << (Ydim/2) - (cropSize/2) + i << " (Xdim/2) - (cropSize/2) + j " << (Xdim/2) - (cropSize/2) + j << std::endl;
+				#endif
```

Comment:
If the code works, I would remove the debuging macros

Reply:
Done